crates/scroll/primitives/src/receipt.rs

```diff
@@ -0,0 +1,281 @@
+    }
+}
+
+impl RlpEncodableReceipt for ScrollReceipt {
+    fn rlp_encoded_length_with_bloom(&self, bloom: &Bloom) -> usize {
+        let mut len = self.eip2718_encoded_length_with_bloom(bloom);
+        if !self.tx_type().is_legacy() {
+            len += Header {
+                list: false,
+                payload_length: self.eip2718_encoded_length_with_bloom(bloom),
+            }
+            .length();
+        }
+
+        len
+    }
+
+    fn rlp_encode_with_bloom(&self, bloom: &Bloom, out: &mut dyn BufMut) {
+        if !self.tx_type().is_legacy() {
+            Header { list: false, payload_length: self.eip2718_encoded_length_with_bloom(bloom) }
+                .encode(out);
+        }
+        self.eip2718_encode_with_bloom(bloom, out);
+    }
+}
+
+impl RlpDecodableReceipt for ScrollReceipt {
+    fn rlp_decode_with_bloom(buf: &mut &[u8]) -> alloy_rlp::Result<ReceiptWithBloom<Self>> {
+        let header_buf = &mut &**buf;
+        let header = Header::decode(header_buf)?;
+
+        // Legacy receipt, reuse initial buffer without advancing
+        if header.list {
+            return Self::rlp_decode_inner(buf, ScrollTxType::Legacy)
+        }
+
+        // Otherwise, advance the buffer and try decoding type flag followed by receipt
+        *buf = *header_buf;
+
+        let remaining = buf.len();
+        let tx_type = ScrollTxType::decode(buf)?;
+        let this = Self::rlp_decode_inner(buf, tx_type)?;
+
+        if buf.len() + header.payload_length != remaining {
+            return Err(alloy_rlp::Error::UnexpectedLength);
+        }
+
+        Ok(this)
+    }
+}
+
+impl TxReceipt for ScrollReceipt {
+    type Log = Log;
+
+    fn status_or_post_state(&self) -> Eip658Value {
+        self.as_receipt().status_or_post_state()
+    }
+
+    fn status(&self) -> bool {
+        self.as_receipt().status()
+    }
+
+    fn bloom(&self) -> Bloom {
+        self.as_receipt().bloom()
+    }
+
+    fn cumulative_gas_used(&self) -> u64 {
+        self.as_receipt().cumulative_gas_used()
+    }
+
+    fn logs(&self) -> &[Log] {
+        self.as_receipt().logs()
+    }
+}
+
+impl Typed2718 for ScrollReceipt {
+    fn ty(&self) -> u8 {
+        self.tx_type().into()
+    }
+}
+
+impl InMemorySize for ScrollReceipt {
+    fn size(&self) -> usize {
+        self.as_receipt().size()
+    }
+}
+
+impl reth_primitives_traits::Receipt for ScrollReceipt {}
+
+#[cfg(feature = "reth-codec")]
+mod compact {
+    use super::*;
+    use alloc::borrow::Cow;
+    use alloy_primitives::U256;
+    use reth_codecs::Compact;
+
+    #[derive(reth_codecs::CompactZstd)]
+    #[reth_zstd(
+        compressor = reth_zstd_compressors::RECEIPT_COMPRESSOR,
+        decompressor = reth_zstd_compressors::RECEIPT_DECOMPRESSOR
+    )]
+    struct CompactScrollReceipt<'a> {
+        tx_type: ScrollTxType,
+        success: bool,
+        cumulative_gas_used: u64,
+        logs: Cow<'a, Vec<Log>>,
+        l1_fee: Option<U256>,
```

Comment:
I used an `Option` here which allows us to save 5 bits of storage for every L1 message receipt. It doesn't add much complexity in the code so I think we can keep it.